File: NodeDesignPatterns/state.ts
let doc=`

                +----------------------+
                |       Context        |
                |----------------------|
                | state: State         |
                |----------------------|
                | request()            |
                +----------+----------+
                           |
                           | delegates
                           v
            +-------------------------------+
            |          State (Interface)    |
            |-------------------------------|
            | handle(context)               |
            +---------------+---------------+
                            |
            +---------------+-------------------+
            |               |                   |
            v               v                   v
    +----------------+ +----------------+ +----------------+
    | ConcreteStateA | | ConcreteStateB | | ConcreteStateC |
    |----------------| |----------------| |----------------|
    | handle()       | | handle()       | | handle()       |
    | (changes state)| | (changes state)| | (changes state)|
    +----------------+ +----------------+ +----------------+


`

console.log(doc)


abstract class State{
    abstract press_button(fan:Fan): void;

}


class Offstate extends State{
    press_button(fan:Fan): void {
        console.log("Fan On")
        fan.state=new Onstate()
    }
}


class Onstate extends State{

    press_button(fan:Fan): void {
        console.log("Fan Off")
        fan.state=new Offstate()
    }

}

class Fan{
    state : State
    constructor(){
        this.state=new Offstate()
    }
    press_button(){
        this.state.press_button(this)
    }
}

const fan = new Fan();

fan.press_button();
fan.press_button();
fan.press_button();

File: PythonDesignPatterns/state.py

"""
                +----------------------+
                |       Context        |
                |----------------------|
                | state: State        |
                |----------------------|
                | request()           |
                +----------+----------+
                           |
                           | delegates
                           v
        +-------------------------------+
        |          State (Interface)    |
        |-------------------------------|
        | handle(context)              |
        +---------------+---------------+
                        |
        +---------------+-------------------+
        |               |                   |
        v               v                   v
+----------------+ +----------------+ +----------------+
| ConcreteStateA | | ConcreteStateB | | ConcreteStateC |
|----------------| |----------------| |----------------|
| handle()       | | handle()       | | handle()       |
| (changes state)| | (changes state)| | (changes state)|
+----------------+ +----------------+ +----------------+


"""

print(__doc__)

from abc import ABC, abstractmethod


class state(ABC):
    @abstractmethod
    def press_button(self,obj):
        pass

class Offstate(state):

    def press_button(self,fan):
        print("Fan turned ON")
        fan.state=Onstate()

class Onstate(state):
    def press_button(self,fan):
        print("Fan turned off")
        fan.state=Offstate()


class Fan:
    def __init__(self):
        self.state=Offstate()

    def press_button(self):
        self.state.press_button(self)


fan = Fan()

fan.press_button()
fan.press_button()
fan.press_button()